File: go/arrays_hashing/0001_two_sum/two_sum_test.go
package twosum

import (
	"reflect"
	"testing"
)

func TestTwoSum(t *testing.T) {
	tests := []struct {
		name   string
		nums   []int
		target int
		want   []int
	}{
		{"Example 1", []int{2, 7, 11, 15}, 9, []int{0, 1}},
		{"Example 2", []int{3, 2, 4}, 6, []int{1, 2}},
		{"Example 3", []int{3, 3}, 6, []int{0, 1}},
		{"No Solution", []int{1, 2, 3}, 7, nil},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := TwoSumHashMap(tt.nums, tt.target); !reflect.DeepEqual(got, tt.want) {
				t.Errorf("TwoSumHashMap() = %v, want %v", got, tt.want)
			}

			if got := TwoSumArray(tt.nums, tt.target); !reflect.DeepEqual(got, tt.want) {
				t.Errorf("TwoSumArray() = %v, want %v", got, tt.want)
			}
		})
	}
}

var benchNums = []int{11, 15, 1, 8, 9, 14, 22, 33, 44, 55, 66, 77, 88, 99, 100, 101, 102, 103, 104, 105, 2, 7}
var benchTarget = 9

func BenchmarkTwoSumHashMap(b *testing.B) {
	for i := 0; i < b.N; i++ {
		TwoSumHashMap(benchNums, benchTarget)
	}
}

func BenchmarkTwoSumArray(b *testing.B) {
	for i := 0; i < b.N; i++ {
		TwoSumArray(benchNums, benchTarget)
	}
}

File: python/arrays_hashing/0001_two_sum/two_sum_test.py
import unittest
import timeit

from two_sum import Solution

class TestTwoSum(unittest.TestCase):
    def setUp(self):
        self.sol = Solution()

    def test_two_sum(self):
        tests = [
            {"name": "Example 1", "nums": [2, 7, 11, 15], "target": 9, "want": [0, 1]},
            {"name": "Example 2", "nums": [3, 2, 4], "target": 6, "want": [1, 2]},
            {"name": "Example 3", "nums": [3, 3], "target": 6, "want": [0, 1]},
            {"name": "No Solution", "nums": [1, 2, 3], "target": 7, "want": None},
        ]

        for tt in tests:
            with self.subTest(name=tt["name"]):
                
                got_dict = self.sol.twoSumDictionary(tt["nums"], tt["target"])
                self.assertEqual(got_dict, tt["want"], 
                                 f"twoSumDictionary() = {got_dict}, want {tt['want']}")

                got_array = self.sol.twoSumArray(tt["nums"], tt["target"])
                self.assertEqual(got_array, tt["want"], 
                                 f"twoSumArray() = {got_array}, want {tt['want']}")

def run_benchmarks():
    print("\n--- Running Benchmarks ---")
    
    setup_code = """
from __main__ import Solution
sol = Solution()
benchNums = [11, 15, 1, 8, 9, 14, 22, 33, 44, 55, 66, 77, 88, 99, 100, 101, 102, 103, 104, 105, 2, 7]
benchTarget = 9
    """
    
    iterations = 100000

    time_dict = timeit.timeit(
        "sol.twoSumDictionary(benchNums, benchTarget)", 
        setup=setup_code, 
        number=iterations
    )
    
    time_array = timeit.timeit(
        "sol.twoSumArray(benchNums, benchTarget)", 
        setup=setup_code, 
        number=iterations
    )

    print(f"Benchmark TwoSumDictionary: {time_dict:.4f} seconds ({iterations} loops)")
    print(f"Benchmark TwoSumArray:      {time_array:.4f} seconds ({iterations} loops)")


if __name__ == '__main__':
    unittest.main(exit=False, verbosity=2)
    
    run_benchmarks()